Feature/dataproc persistent logging

jaketf opened the pull request:
This PR is an example of writing logs from an ephemeral cluster to GCS and using a separate single node cluster to look at Spark and YARN History UIs. This is great reusable demo material.

Future Work: 
- Terraform for history server, possibly refactor to be GKE instead of single node cluster pending @yhqs540  suggestion.

jaketf commented:
@yhqs540 can you review this please?

jaketf commented on examples/dataproc-persistent-history-server/terraform/network.tf:
```diff
@@ -0,0 +1,4 @@
+resource "google_compute_network" "hadoop-network" {
```
Should this and subnet.tf be refactored to use [this cloud foundation toolkit module](https://github.com/terraform-google-modules/terraform-google-network/blob/master/README.md)?

morgante replied:
Yes, that is preferred.

jaketf commented on examples/dataproc-persistent-history-server/README.md:
```diff
@@ -0,0 +1,104 @@
+1.  Replace `PROJECT` with your GCP project id in each file.
+1.  Replace `HISTORY_BUCKET` with your GCS bucket for logs in each file.
+1.  Replace `REGION` with your desired GCP Compute region.
+1.  Manually create the `spark-events` folder in your history bucket.
```
Is there a kosher way to create an empty "directory" in a gcs bucket in terraform? Would a 0B object of the same name work? The way spark logging works it needs this directory to already exist.

morgante replied:
Is there any harm in having a file in the folder? Ex. an empty `.keep` file?

jaketf replied:
I think that will be fine I'll run a test.

morgante commented on examples/dataproc-persistent-history-server/README.md:
````diff
@@ -0,0 +1,104 @@
+
+```
+cd workflow_templates
+sed -i 's/PROJCET/your-gcp-project-id/g' *
````
```suggestion
sed -i 's/PROJECT/your-gcp-project-id/g' *
```

morgante commented on examples/dataproc-persistent-history-server/README.md:
```diff
@@ -0,0 +1,104 @@
+sed -i 's/REGION/us-central1/g' *
+
+cd cluster_templates
+sed -i 's/PROJCET/your-gcp-project-id/g' *
```
```suggestion
sed -i 's/PROJECT/your-gcp-project-id/g' *
```

morgante commented on examples/dataproc-persistent-history-server/terraform/firewall.tf:
```diff
@@ -0,0 +1,44 @@
+resource "google_compute_firewall" "history-ui-access" {
+    name = "${var.network}-hadoop-history-ui-access"
```
Please run `terraform fmt`.

morgante commented on examples/dataproc-persistent-history-server/terraform/variables.tf:
```diff
@@ -0,0 +1,32 @@
+}
+
+variable "network" {
+    description = "The network that your hadoop clusters should use"
```
Use or create? "Use" implies an existing network, but this code actually creates it.

jaketf commented:
@morgante this is ready for re-review. Apologies for the delay and commit-sprawl.

morgante commented on examples/dataproc-persistent-history-server/terraform/long-running-cluster.tf:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    initialization_action = {
+      script = "gs://${var.history-bucket}/init_actions/disable_history_servers.sh"
```
How about referencing google_storage_bucket_object here to avoid the implicit dependency?

morgante commented on examples/dataproc-persistent-history-server/terraform/service-account.tf:
```diff
@@ -0,0 +1,27 @@
+  project = "${var.project}"
+  role    = "roles/dataproc.worker"
+
+  member = "serviceAccount:history-server-account@${var.project}.iam.gserviceaccount.com"
```
```suggestion
  member = "${google_service_account.history-server-account.email}"
```

jaketf replied:
Done. FYI needed to have `serviceAccount:` prefix.

morgante commented on examples/dataproc-persistent-history-server/terraform/service-account.tf:
```diff
@@ -0,0 +1,27 @@
+ * limitations under the License.
+ */
+
+resource "google_service_account" "history-server-account" {
```
Is "server-account" (vs. "service-account") intentional?

jaketf replied:
It's the account for the history server.

jaketf commented on examples/dataproc-persistent-history-server/README.md:
````diff
@@ -0,0 +1,117 @@
+
+```
+touch .keep
+gsutil cp .keep gs://your-history-bucket/spark-events/.keep
````
I want to provide a gcloud route and a Terraform route. Some people will only wanna spin parts of this up and don't have terraform installed.

jaketf replied:
I added a section to the README to clarify this. Let me know if that suffices.

morgante replied:
Got it, LGTM.

jaketf commented on examples/dataproc-persistent-history-server/README.md:
````diff
@@ -0,0 +1,117 @@
+Create the history server.
+
+```sh
+gcloud beta dataproc clusters import \
````
We already create the history server in terraform. This history server is stateless (all the logs live on GCS) so theoretically this allows someone to spin it up only when they want to look at the job history and then tear it down.

jaketf commented on examples/dataproc-persistent-history-server/terraform/firewall.tf:
```diff
@@ -0,0 +1,57 @@
+
+  allow = {
+    protocol = "udp"
+    ports    = ["0-65535"]
```
I followed the guidance [here](https://cloud.google.com/dataproc/docs/concepts/configuring-clusters/network). The cluster nodes need a lot open to communicate with one another. TBH I'm not sure which ports exactly. @yhqs540 do you know of an exhaustive list of ports needed for hadoop communication?

jaketf replied:
confirmed with @yhqs540 offline, it's unrealistic to restrict ports because of how many ports different hadoop services use to communicate.

morgante replied:
Got it.

jaketf commented on examples/dataproc-persistent-history-server/terraform/history-server.tf:
```diff
@@ -0,0 +1,58 @@
+ */
+
+resource "google_dataproc_cluster" "history-server" {
+  depends_on = ["google_storage_bucket.history-bucket",
```
`terraform fmt` didn't change this section. I moved the first item of the list to a new line from the bracket is that what you were looking for?

jaketf commented:
@morgante addressed your comments regarding referencing things explicitly instead of over using variables and depends on. See changes to read me if that clears up why I duplicate things between terraform and gcloud.

morgante commented on examples/dataproc-persistent-history-server/terraform/history-bucket.tf:
```diff
@@ -0,0 +1,62 @@
+resource "google_storage_bucket_object" "history-server-yaml" {
+  bucket       = "${google_storage_bucket.history-bucket.name}"
+  name         = "init_actions/disable_history_servers.sh"
+  name         = "cluster_templates/history-server.yaml"
```
`name` looks duplicated.
